### ISSUE TITLE
What to do with virtual dotvvm properties?

### ISSUE THREAD

**Issue by exyi** [riganti/dotvvm#802]
We have a concept of virtual dotvvm properties that create `DotvvmProperty` in the background from a standard C# property when it is annotated with `MarkupOptionsAttribute`. It has some obvious limitations (the property can not contain bindings, can not be inherited) and some less obvious limitations (can not contain even `resource` binding, can not be set from styles).

Basically, the same applies to property groups, which have two modes - is set directly to a property (for example [`HtmlGenericControl`](https://github.com/riganti/dotvvm/blob/master/src/DotVVM.Framework/Controls/HtmlGenericControl.cs#L56)) or it's stored in `control.properties` and accessed through `VirtualPropertyGroupDictionary` (for example [`RouteLink.Params-...`](https://github.com/riganti/dotvvm/blob/master/src/DotVVM.Framework/Controls/RouteLink.cs)) (in this case, virtual word is very overloaded and it means exactly the opposite to the paragraph above :weary:)

Since this feature showed itself as not very useful (AFAIK it's actually not used), we could maybe drop it for the sake of consistency in the following major release.

Or, we could at least fix the inconsistencies - allow setting these properties though compile-time style. We could also allow them to hold bindings by a `ValueOrBinding<T>` type (see #801). The advantage of these properties is that they are declared with a lot less boilerplate, and maybe this could make writing DotVVM controls less scary.

**Comment by MichalTichy** [comment]
The Class-_ClassName_ attribute on HtmlGenericControl is also VirtualProperty, Isn't it?

Because I've seen this used on many places.
But that's probably the only type (except of those on RouteLienk) of Virtual properties I've ever seen used. Unfortunately there is not reference to virtual properties anywhere in DotVVM doc :( .

**Comment by exyi** [comment]
> The Class-ClassName attribute on HtmlGenericControl is also VirtualProperty, Isn't it?

Not the one I mean. I mean properties that are not backed in the `DotvvmControl.properties` field, which `Class-X` are, while `HtmlGenericControl.Attributes` are not, for example

> Unfortunately there is not reference to virtual properties anywhere in DotVVM doc :( .

That's because it's a hidden feature I'd find comfortable removing by now :dancers:

**Comment by MichalTichy** [comment]
I've done a search for "html:" across all the DotVVM project I've access to, and found only 2 projects which ever used it (21 occurrences total across 4 files).

So I think that we can kill this feature without anyone noticing.

**Comment by MichalTichy** [comment]
The vote for killing it only applies if I get correctly that this would not affect RouteLink Param- , ... properties on RouteLink. Usage of those is entirely different story.

**Comment by exyi** [comment]
> I've done a search for "html:" across all the DotVVM projects I've access to, and found only 2 projects which ever used it (21 occurrences total across 4 files).

No, Dothtml markup is unrelated to this issue. The problem is that users can declare dotvvm properties like this:


```csharp
[MarkupOptions]
public bool X { get; set; }
```

without the terrible boilerplate. Almost nobody does that, but it's still kinda supported. Other thing is, that it might be quite cool, since it makes the property declaration less painful.
